### PR TITLE
zsh completion

### PR DESCRIPTION
**Pull request by crpb**
zsh completion

### CODE REVIEW

**Comment by jwilk** [comment]
I have no idea how to test this.

**Comment by crpb** [comment]
maybe there is an easier way but well.. 
```
sudo apt-get install zsh
sudo useradd -m -s /bin/zsh test
sudo -i -u test
# hit "2" (auto populate defaults)
git clone --branch completion https://github.com/crpb/ubanner
chmod +x ubanner/ubanner
sed -i '2i PATH="$HOME/ubanner:$PATH"\nfpath=($fpath ~/ubanner/completion)' ~/.zshrc
source ~/.zshrc
ubanner <TABTABTAB>
```
or something alike o_0
```

                    ░░▒███████████████▒░░                                       ░▒▒███████████████▒▒░                                          ░░▒██████████████▒░░
                 ░▒███████████████████████▒                                  ░▒████████████████████████▒                                     ▒███████████████████████░
               ░████████████████████████████▒                               █████████████████████████████▒                                 ▒███████████████████████████░
          ▒██▒▒███████████████████████████████                            ░████████████████████████████████                               ██████▒██████████▒▒▒██████████░
       ░██▒ ░█▒▒██░      ░▒███████▒       ▒█████▒██░                     ░████▒░    ░▒██████████░     ░█████                             ▒███▒░▒███▒░  ▒████▒░  ░▒███████
      ▒█▒ ░██  ▒█░         ░████▒          ██  ▒░ ░██░             ░░▒▒░░████          ▒██████░         ░███▒░░▒▒░                 ▒███████████▒░   ░▒████▒░▒██▒    ░█████████▒░
     ██░ ▒░█  ░█▒   ░░░░     ██▒     ░░░   ░█▒  ▒█▒ ░██          ▒██▒░▒█████             ████            ▒█████░▒███             ▒█▒░  ▒█████░    ░█▒░  ▒██▒░  ░▒      ▒███   ▒██
    ██    ░   ██  ▒█▒████░    ░    ██▒████  ▒█░  ▒ ░  ██░       ██▒    ▒███▒  ░▒████▒     █▒    ░█████▒   ████    ░██░          ▒█▒     ███░      ░ ░▒▒█▒░▒████░         ▒█     ██
   ██        ▒██  ████████        ░███████▒  ██        ██      ▒█▒      ███▒  █▒▒█████         ░█▒▒█████  ███░      ██░         ██     ▒█░       ░▒██▒▒▒███░░  ░▒▒        ░█▒   ░█▒
   ██       ░███  ████░ ░░        ░████  █░ ░██▒       ▒█▒     ██       ████ ░█████▒▒█         ▒█████▒▒█  ███       ░█▒         ██   ░█▒             ░▒▒█░ ░▒██▒░           ▒█░ ░██
   ██       ▒███▒  ▒██▒▒░  ░░░▒░░   ▒▒▒░▒░  ████       ░█▒     ██       ████  ▒███  ▒▒          ████ ░█░ ▒███        ██         ██  ▒█░          ░▒▒████▒██▒▒░ ░▒█▒          ░██▒█▒
   ░█▒       █████      ░▒▒░░░░░░░▒▒      ░█████       ██░     ██       █████   ░░░░  ░▒▒███▒▒▒░ ░░░░░  ░████       ░█▒         ▒████          ▒█▒ ░░       ░▒▒   ▒█▒          ███
    ░██      ▒██░        ░▒      ░▒░        ███░      ▒█░      ▒█▒      █████▒  ░░ ▒███▒      ░████ ░░ ░█████       ██░          ███         ░██   ░██      ██▒    ░██░         ██░
     ░██░     ▒█▒       ▒█▒       ██░       ██▒      ██░        ██▒     ████   ▒█▒███ ░██    ▒█▒ ▒██▒██  ▒███     ░██░           ██         ▒█▒    ░░        ░▒      ▒█░        ░██
      ░██░     ██▒                         ▒█▒     ░██░          ▒██▒░░▒███    ░█▒ ░████▒█▒░█▒  ▒█▒ ░█▒   ███▒░░░███░           ██░        ▒█▒                        ▒█░        ██▒
       ▒██      ██                        ░██     ░██░             ░▒▒▒████  ░████▒ ▒█░ ░███   ▒█░ ░█████ █████▒▒▒░            ▒█▒        ░█▒                          ██        ░██
       ███      ██░          ░▒▒          ▒█▒     ░██                  ▒███░ ██░░▒███░ ░██░   ██  ▒█▒░ ██ ████                 ██         ██░          ▒███           ███▒        ▒█▒
      ░██░      ██         ░████▒         ▒█░      ██▒                 ░████  ▒█░▒██░ ░█▒    █▒ ░█▒  ░██ ▒███▒                ▒█▒        ▒███░        █████         ░█████         ██
      ███       ██▒        █████          ██▒      ░██                  █████  ███▒   ░        ▒█░  ███ ░████                 ██         ██████░      ▒██▒         ▒██████▒        ██▒
     ░██       █████░     ░░▒▒░         ▒████       ▒██                  ████████░                ░██▒██████░                ▒██        ░██▒█████▒░  ▒▒▒▒       ░█████▒ ░██        ░██
     ██░      ░███████▒░  ░▒▒▒░      ░▒██████▒       ██░                  █████░                 ▒█▒   ▒███░                 ███        ██░  ▒███████▒▒░░░░▒▒███████░    ██░       ░██░
    ███       ██  ▒████████▒▒▒▒▒▒▒██████▒░  ██░      ░██                 ░███░                 ▒██░      ▒██░                ███        ██     ░▒███████████████▒░       ▒█▒       ░██░
   ░██       ██░    ░▒███████████████▒░     ░██       ███              ▒███░                 ▒██▒          ▒██▒              ██▒       ▒██          ░▒▒▒▒▒▒▒░░           ░██       ░██░
   ▒▒░      ░▒▒          ░░▒▒▒▒▒▒░░          ▒▒▒       ▒▒           ░███▒░                ▒███▒              ▒███░           ▒▒░       ▒▒░                                ▒▒       ░▒▒
                                                                 ░▒███░               ░▒██████▒░               ░███▒░
                                                               ░▒▒▒░                ░▒▒▒░    ░▒▒▒░                ░▒▒▒░

```

**Comment by jwilk** [comment]
I've simpified zsh test setup to this:
```shell
tmpdir=$(mktemp -d)
cat > "$tmpdir/.zshrc" <<EOF
path+=(/path/to/ubanner)
fpath+=(/path/to/ubanner/completion)
autoload -Uz compinit
compinit
EOF
ZDOTDIR=$tmpdir zsh
```

File: completion/_ubanner [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+# zsh completion for ubanner
+
+_arguments -s -S \
+  {-h,--help}'[show help]' \
+  '--version[show version information and exit]' \
+  '--list-fonts[print list of available font faces]' \
+  '--trim[trim leading/trailing empty lines]' \
+  {-f,--font}'[use this fontface]:fontname:->fonts' \
+  {-s,--font-size}'[font size in pixels (default: 28)]:size:(28 36 42 50 58 64 72 80 88 96)' \
```
Comment (jwilk) [inline completion/_ubanner]: That's a weird list of integers. Sometimes the distance between them is 6, sometimes 8, with no apparent pattern.
Anyway: typing 64 is easier than 6[tab], so just get rid of the list?

File: completion/_ubanner [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  {-S,--full-screen}'[full screen mode]'
```
Comment (jwilk) [inline completion/_ubanner]: Please steal the description from the help message.

**Review by jwilk — CHANGES_REQUESTED** [review]

File: completion/_ubanner
Comment (jwilk) [inline completion/_ubanner:1]: The directory name should indicate that is for zsh. Maybe `completion/zsh/` or `zshcompletion/`?

Comment (jwilk) [inline completion/_ubanner:1]: Please add copyright statement and a license info.
For example, in `Makefile` it looks like this:
https://github.com/jwilk/ubanner/blob/5602b78d98289ca1576cd4019976ab751738f162/Makefile#L1-L2

**Review by jwilk — CHANGES_REQUESTED** [review]

File: completion/_ubanner (hunk 1 above)
Comment (crpb) [reply completion/_ubanner]: > Anyway: typing 64 is easier than 6[tab], so just get rid of the list?
that is actually there so you can complete them like this
```
% ubanner --font-size
completing size
28  36  42  50  58  64  72  80  88  96
```
i just started with a number and kept going ^_^

File: completion/_ubanner
Comment (crpb) [reply completion/_ubanner:1]: i guess `completion/$shell` make the most sense.
but if it's only zsh and bash than they can be discerned by the leading underscore fore zsh :stuck_out_tongue_winking_eye:

Comment (crpb) [reply completion/_ubanner:1]: i don't think i ever saw a completion file with a license statement :astonished: and now after searching its only a [few](https://github.com/search?q=repo%3Azsh-users%2Fzsh++path%3ACompletion++%2F%5E%23.*%5C+mit%2F+OR+%2F%5E%23.*gnu%2F+OR+%2F%5E%23.*license%2F+&type=code) if that 'search' was working...
when you look at the big [picture](https://github.com/search?q=repo%3Azsh-users%2Fzsh++path%3ACompletion+&type=code) it seems nobody cares :heart_eyes_cat:

File: completion/zsh [hunk 3]
```diff
@@ -0,0 +1,24 @@
+#compdef ubanner
+# Copyright © 2024 Christopher Bock <christopher@bocki.com>
+# # SPDX-License-Identifier: MIT
+local arguments
+
+arguments=(
+  '--trim[trim leading/trailing empty lines]'
+  '(-f --font)'{-f,--font}'[use this fontface]:font:->fonts'
```
Comment (jwilk) [inline completion/zsh]:
```suggestion
  '(-f --font)'{-f,--font}'[use this font face]:font:->fonts'
```

File: completion/zsh [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#compdef ubanner
+# Copyright © 2024 Christopher Bock <christopher@bocki.com>
+# # SPDX-License-Identifier: MIT
```
Comment (jwilk) [inline completion/zsh]:
```suggestion

# Copyright © 2024 Christopher Bock <christopher@bocki.com>
# SPDX-License-Identifier: MIT

```

File: completion/_ubanner
Comment (jwilk) [reply completion/_ubanner:1]: Now there's no underscore, so I can't test it without jumping through an extra hop.
Can you make it `completion/zsh/_ubanner`?

File: completion/zsh [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+# Copyright © 2024 Christopher Bock <christopher@bocki.com>
+# SPDX-License-Identifier: MIT
+
+local arguments
+
+arguments=(
+  '--trim[trim leading/trailing empty lines]'
+  '(-f --font)'{-f,--font}'[use this font face]:font:->fonts'
+  + '(fs)'
+  '(-s --font-size)'{-s,--font-size}'[font size in pixels]: :_numbers -d28 -l14 -m280'
```
Comment (jwilk) [inline completion/zsh]: What are the `_numbers` options for? They don't seem to have any effect. In particular, I'm not being shown what's the default for `--font-size`. :confused:

```console
% ubanner -⭾
…
--font-size    -s  -- font size in pixels                                     
…
```

File: completion/zsh [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  '(-S --full-screen)'{-S,--full-screen}'[max font size to fit terminal window]'
+  + '(excl)'
+  {-h,--help}:'[show help]'
+  '--version:[show version information and exit]'
+  '--list-fonts:[print list of available font faces]'
```
Comment (jwilk) [inline completion/zsh]: I'm not being shown descriptions for these options. :confused:
```console
% ubanner -⭾
…          
--help             --list-fonts       --version          -h
```

Comment (jwilk) [reply completion/zsh]:
```suggestion
  + '(excl)'
  {-h,--help}'[show help]'
  '--version[show version information and exit]'
  '--list-fonts[print list of available font faces]'
```

File: completion/zsh (hunk 5 above)
Comment (crpb) [reply completion/zsh]: Now after using
```
zstyle ':completion:*' menu select=1
zstyle ':completion:*' completer _complete
zstyle ':completion:*:descriptions'    format $'%{\e[0;31m%}completing %B%d%b%{\e[0m%}'
zstyle ':completion:*:messages'        format '%d'
zstyle ':completion:*:options'         auto-description '%d'
```
it should be clear?

**Review by crpb — COMMENTED** [review]
🦆🐔🪱

**Pull request merged**